### ISSUE TITLE
AssertionError: Items are not equal: ACTUAL: 2147483647 DESIRED: -2147483648

### ISSUE THREAD

**Issue by sandipmgiri** [tensorflow/tensorflow#9216]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  
Ubuntu 16.04 (ppc64le)
- **TensorFlow installed from (source or binary)**: 
 Installed from source (v1.0.1)
- **TensorFlow version (use command below)**:    
('v1.0.1-0-ge895d5c-dirty', '1.0.1')
- **Bazel version (if compiling from source)**: 
bazel release 0.4.4-2017-04-10 (@80a07b5)
- **CUDA/cuDNN version**: 
In disable mode
- **Exact command to reproduce**:
bazel test //tensorflow/python/kernel_tests:cast_op_test

### Describe the problem
Built TF successfully , however I am getting `Items are not equal` error while running the `cast_op_test` 

To cross verify the test results , I ran this test on X86 vm and that passed successfully. This test is failing only on ppc64le platform . Here I would like to know your suggestions and comments. 

### Source code / logs
````
$ bazel test //tensorflow/python/kernel_tests:cast_op_test


exec ${PAGER:-/usr/bin/less} "$0" || exit 1
-----------------------------------------------------------------------------
I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 16 visible devices
I tensorflow/compiler/xla/service/service.cc:180] XLA service executing computations on platform Host. Devices:
I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (0): <undefined>, <undefined>
/root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/cast_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/cast_op_test.py:62: ComplexWarning: Casting complex values to real discards the imaginary part
  np_ans = x.astype(dtype)
....F.W tensorflow/core/framework/op_kernel.cc:983] Unimplemented: Cast int64 to string is not supported
E tensorflow/core/common_runtime/executor.cc:594] Executor failed to create kernel. Unimplemented: Cast int64 to string is not supported
         [[Node: Cast = Cast[DstT=DT_STRING, SrcT=DT_INT64, _device="/job:localhost/replica:0/task:0/cpu:0"](Cast/x)]]
........
======================================================================
FAIL: testInfNan (__main__.CastOpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/cast_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/cast_op_test.py", line 150, in testInfNan
    self._compare(np.inf, np.int32, i4.min, False)
  File "/root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/cast_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/cast_op_test.py", line 124, in _compare
    x, dst_dtype, use_gpu=use_gpu), dst_dtype(expected))
  File "/usr/lib64/python2.7/site-packages/numpy/testing/utils.py", line 425, in assert_equal
    raise AssertionError(msg)
AssertionError:
Items are not equal:
 ACTUAL: 2147483647
 DESIRED: -2147483648

----------------------------------------------------------------------
Ran 14 tests in 2.485s

FAILED (failures=1)
`

```

**Comment by asimshankar** [comment]
Thanks for pointing this out.

Support for big-endian architecture was added by @namrata-ibm in commit 1e44b15ff
Support for big-endian architectures in TensorFlow is community driven and we don't test that as part of every release at this time.

@namrata-ibm : Could you take a look? Seems like this is the line that is failing:
https://github.com/tensorflow/tensorflow/blob/87cdfafd44ff5e332fd820608783432fea83a4c9/tensorflow/python/kernel_tests/cast_op_test.py#L147

**Comment by asimshankar** [comment]
Oh, I also notice that you mentioned you're running on `ppc64le`.
Isn't that little endian? If so, could you share the output of the following python script:

```
import sys
print sys.byteorder
```

Because the test failure suggests that `sys.byteorder` is `big` even though your architecture is little-endian?

**Comment by namrata-ibm** [comment]
The failure seems to be at line number 150 from above log:

```
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/68a62076e91007a7908bc42a32e4cff9/execroot/tensorflow/bazel-out/local-opt/bin/tensorflow/python/kernel_tests/cast_op_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/cast_op_test.py", line 150, in testInfNan

```

in function:

`
    self._compare(np.inf, np.int32, i4.min, False)
`

@asimshankar ,

ppc64le being little endian, doesn't enter the condition for `big endian` and follows the original code just like x86.

I can also see error : 
` E tensorflow/core/common_runtime/executor.cc:594] Executor failed to create kernel.  Unimplemented: Cast int64 to string is not supported` in above log.
This would need some debugging on ppc64le.

**Comment by asimshankar** [comment]
Oh, apologies, dunno why I misread the line.

@sandipmgiri : Does seem that the cast kernel (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/cast_op.cc#L116) isn't quite doing what we expected on ppc64le.

Are you able to trace that? Contributions to fix it will be welcome.

**Comment by sandipmgiri** [comment]
Yes, the failure at line number 150 i.e. self._compare(np.inf, np.int32, i4.min, False)

To understand the issue, I ran this test in standalone file, as below: 

```
def _cast(self, x, dtype, use_gpu=False):
    with self.test_session(use_gpu=use_gpu):
      val = constant_op.constant(x, self._toDataType(np.array([x]).dtype))
      print("---The value of VAL argument is ", val)
      print("---The value of self._toDataType(dtype) arg is ", self._toDataType(dtype))
      return math_ops.cast(val, self._toDataType(dtype), name="cast").eval()

 # Special values like int32max, int64min, inf, -inf, nan casted to
  # integer values in somewhat unexpected ways. And they behave
  # differently on CPU and GPU.
  def _compare(self, x, dst_dtype, expected, use_gpu=False):
    np.testing.assert_equal(
        self._cast(
            x, dst_dtype, use_gpu=use_gpu), dst_dtype(expected))

  def testInfNan(self):
      i4 = np.iinfo(np.int32)
      self._compare(np.inf, np.int32, i4.min, False)
```

after that I got to know that the actual failure at line :  
`return math_ops.cast(val, self._toDataType(dtype), name="cast").eval()`

return value on X86 =  `-2147483648`
return value on ppc64le = `2147483647`

Here 3 arguments passed to math_ops.cast function , and in our test the value of each argument as below (I have printed these values in my code):
val = Tensor("Const:0", shape=(), dtype=float64, device=/device:CPU:0)
self._toDataType(dtype) = <dtype: 'int32'>
name = "cast"

Now here I am not able to understand the flow for math_ops.cast function:
I had written 4 to 6 print statements in math_ops.py file but not even a single print statement was executed.

Looks like the main problem in math.ops.cast function (for ppc64le).  Any comments would appreciate . 

Thanks!

**Comment by sandipmgiri** [comment]
Further analysis report :
Here , `math_ops.cast(val, self._toDataType(dtype), name="cast").eval()`  function producing the different results on X86 and PPC64LE.

The first part from above function i.e.` math_ops.cast(val, self._toDataType(dtype), name="cast")`  generating same output on both the platform = ` Tensor("cast:0", shape=(), dtype=int32, device=/device:CPU:0)`

After that performing eval() operation on ` Tensor("cast:0", shape=(), dtype=int32, device=/device:CPU:0)` and results are :
on x86 =` -2147483648`
on ppc64le = `2147483647`

The main problem  in  eval() function (i.e. in 2nd part). I would like to know more about eval() function. How that eval function is working here ? Why that function behavior is different on each platform ?

**Comment by sandipmgiri** [comment]
Yes, the actual failure at line number 150 , in function `self._compare(np.inf, np.int32, i4.min, False)`

For little-endian, I tried changing the value in above function :
This test passed on both the platform (x86 and ppc64le) after replacing the `np.inf` with `-np.inf`   
now getting same results  i.e. `-2147483648`

@asimshankar , Is there need to take care of ?

**Comment by jjhelmus** [comment]
I've also run into this issue on a `ppc64le` system.

Casting a non-finite float point values (inf, -inf, nan) to an integer is an undefined behavior in C (see section 6.3.1.4 of the C99 standards).  These tests seems to be expecting a specific result from this undefined behavior which is not portable.  The expected result, the largest negative number supported by integer type being cast to, happens to be the integer indefinite value which the x87 FPU gives for floating-point invalid-operations.

I'm not certain how these tests can be made portable given that the behavior being exercised is  architecture specific.

**Comment by sandipmgiri** [comment]
Closing as duplicate of https://github.com/tensorflow/tensorflow/issues/9360, see discussion in https://github.com/tensorflow/tensorflow/issues/9360.

**Comment by sandipmgiri** [comment]
This issue resolved in PR #10522 (merged commit d0d2308)
